airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/testutils/TestDatabase.java

```diff
@@ -0,0 +1,293 @@
+    return self();
+  }
+
+  /**
+   * Executes SQL statements as root to provide the necessary isolation for the lifetime of this
+   * object. This typically entails at least a CREATE DATABASE and a CREATE USER. Also Initializes the
+   * {@link DataSource} and {@link DSLContext} owned by this object.
+   */
+  final public T initialized() {
+    inContainerBootstrapCmd().forEach(this::execInContainer);
+    this.dataSource = DataSourceFactory.create(
+        getUserName(),
+        getPassword(),
+        getDatabaseDriver().getDriverClassName(),
+        getJdbcUrl(),
+        connectionProperties);
+    this.dslContext = DSLContextFactory.create(dataSource, getSqlDialect());
+    return self();
+  }
+
+  final public boolean isInitialized() {
+    return dslContext != null;
+  }
+
+  abstract protected Stream<Stream<String>> inContainerBootstrapCmd();
+
+  abstract protected Stream<String> inContainerUndoBootstrapCmd();
+
+  abstract public DatabaseDriver getDatabaseDriver();
+
+  abstract public SQLDialect getSqlDialect();
+
+  final public C getContainer() {
+    return container;
+  }
+
+  public String withNamespace(String name) {
+    return name + suffix;
+  }
+
+  public String getDatabaseName() {
+    return withNamespace("db");
+  }
+
+  public String getUserName() {
+    return withNamespace("user");
+  }
+
+  public String getPassword() {
+    return "password";
+  }
+
+  public DataSource getDataSource() {
+    if (!isInitialized()) {
+      throw new RuntimeException("TestDatabase instance is not yet initialized");
+    }
+    return dataSource;
+  }
+
+  final public DSLContext getDslContext() {
+    if (!isInitialized()) {
+      throw new RuntimeException("TestDatabase instance is not yet initialized");
+    }
+    return dslContext;
+  }
+
+  public String getJdbcUrl() {
+    return String.format(
+        getDatabaseDriver().getUrlFormatString(),
+        getContainer().getHost(),
+        getContainer().getFirstMappedPort(),
+        getDatabaseName());
+  }
+
+  public Database getDatabase() {
+    return new Database(getDslContext());
+  }
+
+  protected void execSQL(Stream<String> sql) {
+    try {
+      getDatabase().query(ctx -> {
+        sql.forEach(ctx::execute);
+        return null;
+      });
+    } catch (SQLException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  protected void execInContainer(Stream<String> cmds) {
+    final List<String> cmd = cmds.toList();
+    if (cmd.isEmpty()) {
+      return;
+    }
+    try {
+      LOGGER.debug("executing {}", Strings.join(cmd, " "));
+      final var exec = getContainer().execInContainer(cmd.toArray(new String[0]));
+      if (exec.getExitCode() == 0) {
+        LOGGER.debug("execution success\nstdout:\n{}\nstderr:\n{}", exec.getStdout(), exec.getStderr());
+      } else {
+        LOGGER.error("execution failure, code {}\nstdout:\n{}\nstderr:\n{}", exec.getExitCode(), exec.getStdout(), exec.getStderr());
```

Comment:
should we throw an exception here, rather than continuing to the next cmd?

Reply:
I thought about this and came to the conclusion that we shouldn't rely on the exit codes too much here. Perhaps this should really be logged at warning rather than at error.